File: src/api/src/infrastructure/Yoma.Core.Infrastructure.Database/Opportunity/Entities/Opportunity.cs
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Yoma.Core.Infrastructure.Database.Core.Entities;
using Yoma.Core.Infrastructure.Database.Entity.Entities;
using Yoma.Core.Infrastructure.Database.Lookups.Entities;
using Yoma.Core.Infrastructure.Database.Opportunity.Entities.Lookups;

namespace Yoma.Core.Infrastructure.Database.Opportunity.Entities
{
    [Table("Opportunity", Schema = "Opportunity")]
    [Index(nameof(Title), IsUnique = true)]
    [Index(nameof(TypeId), nameof(OrganizationId), nameof(ZltoReward), nameof(DifficultyId), nameof(CommitmentIntervalId), nameof(CommitmentIntervalCount), nameof(StatusId), nameof(Keywords),
        nameof(DateStart), nameof(DateEnd), nameof(CredentialIssuanceEnabled), nameof(DateCreated), nameof(DateModified))]
    public class Opportunity : BaseEntity<Guid>
    {
        //support specials characters like emojis  
        [Required]
        [Column(TypeName = "nvarchar(255)")]
        public string Title { get; set; }

        //support specials characters like emojis  
        [Required]
        [Column(TypeName = "nvarchar(MAX)")]
        public string Description { get; set; }

        [Required]
        [ForeignKey("TypeId")]
        public Guid TypeId { get; set; }
        public OpportunityType Type { get; set; }

        [Required]
        [ForeignKey("OrganizationId")]
        public Guid OrganizationId { get; set; }
        public Organization Organization { get; set; }

        //support specials characters like emojis  
        [Column(TypeName = "nvarchar(500)")]
        public string? Summary { get; set; }

        //support specials characters like emojis  
        [Column(TypeName = "nvarchar(MAX)")]
        public string? Instructions { get; set; }

        [Column(TypeName = "varchar(2048)")]
        public string? URL { get; set; }

        [Column(TypeName = "decimal(8,2)")]
        public decimal? ZltoReward { get; set; }

        [Column(TypeName = "decimal(12,2)")]
        public decimal? ZltoRewardPool { get; set; }

        [Column(TypeName = "decimal(12,2)")]
        public decimal? ZltoRewardCumulative { get; set; }

        [Column(TypeName = "decimal(8,2)")]
        public decimal? YomaReward { get; set; }

        [Column(TypeName = "decimal(12,2)")]
        public decimal? YomaRewardPool { get; set; }

        [Column(TypeName = "decimal(12,2)")]
        public decimal? YomaRewardCumulative { get; set; }

        [Required]
        public bool VerificationEnabled { get; set; }

        [Column(TypeName = "varchar(20)")]
        public string? VerificationMethod { get; set; }

        [Required]
        [ForeignKey("DifficultyId")]
        public Guid DifficultyId { get; set; }
        public OpportunityDifficulty Difficulty { get; set; }

        [Required]
        [ForeignKey("CommitmentIntervalId")]
        public Guid CommitmentIntervalId { get; set; }
        public TimeInterval CommitmentInterval { get; set; }

        [Required]
        public short CommitmentIntervalCount { get; set; }

        public int? ParticipantLimit { get; set; }

        public int? ParticipantCount { get; set; }

        [Required]
        [ForeignKey("StatusId")]
        public Guid StatusId { get; set; }
        public OpportunityStatus Status { get; set; }

        [Column(TypeName = "varchar(500)")]
        public string? Keywords { get; set; }

        [Required]
        public DateTimeOffset DateStart { get; set; }

        public DateTimeOffset? DateEnd { get; set; }

        [Required]
        public bool CredentialIssuanceEnabled { get; set; }

        [Column(TypeName = "varchar(255)")]
        public string? SSISchemaName { get; set; }

        [Required]
        public DateTimeOffset DateCreated { get; set; }

        [Required]
        [Column(TypeName = "varchar(320)")]
        public string CreatedBy { get; set; }

        [Required]
        public DateTimeOffset DateModified { get; set; }

        [Required]
        [Column(TypeName = "varchar(320)")]
        public string ModifiedBy { get; set; }

        public ICollection<OpportunityCategory> Categories { get; set; }

        public ICollection<OpportunityCountry> Countries { get; set; }

        public ICollection<OpportunityLanguage> Languages { get; set; }

        public ICollection<OpportunitySkill>? Skills { get; set; }

        public ICollection<OpportunityVerificationType>? VerificationTypes { get; set; }
    }
}

File: src/web/youth/src/api/models/opportunity.ts
import type { PaginationFilter } from "./common";
import type { Country, Language, Skill } from "./lookups";

export interface OpportunitySearchFilterAdmin
  extends OpportunitySearchFilterBase {
  startDate: string | null;
  endDate: string | null;
  organizations: string[] | null;
  statuses: Status[] | null;
}

export interface OpportunitySearchResults extends OpportunitySearchResultsBase {
  items: Opportunity[];
}

export interface OpportunitySearchResultsBase {
  totalCount: number | null;
}

export interface Opportunity {
  id: string;
  title: string;
  description: string;
  typeId: string;
  type: string;
  organizationId: string;
  organization: string;
  organizationStatusId: string;
  organizationStatus: OrganizationStatus;
  summary: string | null;
  instructions: string | null;
  url: string | null;
  zltoReward: number | null;
  zltoRewardPool: number | null;
  zltoRewardCumulative: number | null;
  yomaReward: number | null;
  yomaRewardPool: number | null;
  yomaRewardCumulative: number | null;
  verificationEnabled: boolean;
  verificationMethod: VerificationMethod | null;
  difficultyId: string;
  difficulty: string;
  commitmentIntervalId: string;
  commitmentInterval: string;
  commitmentIntervalCount: number | null;
  participantLimit: number | null;
  participantCount: number | null;
  statusId: string;
  status: Status;
  keywords: string[] | null;
  dateStart: string;
  dateEnd: string | null;
  credentialIssuanceEnabled: boolean;
  ssiSchemaName: string | null;
  dateCreated: string;
  createdBy: string;
  dateModified: string;
  modifiedBy: string;
  published: boolean;
  categories: OpportunityCategory[] | null;
  countries: Country[] | null;
  languages: Language[] | null;
  skills: Skill[] | null;
  verificationTypes: OpportunityVerificationType[] | null;
}

export interface OpportunitySearchFilterBase extends PaginationFilter {
  types: string[] | null;
  categories: string[] | null;
  languages: string[] | null;
  countries: string[] | null;
  valueContains: string | null;
}

export interface OpportunityInfo {
  id: string;
  title: string;
  description: string;
  type: string;
  organization: string;
  instructions: string | null;
  uRL: string | null;
  zltoReward: number | null;
  yomaReward: number | null;
  difficulty: string;
  commitmentInterval: string;
  commitmentIntervalCount: number | null;
  participantLimit: number | null;
  participantCountVerificationCompleted: number;
  participantCountVerificationPending: number;
  participantCountTotal: number;
  keywords: string[] | null;
  dateStart: string;
  dateEnd: string | null;
  published: boolean;
  categories: OpportunityCategory[] | null;
  countries: Country[] | null;
  languages: Language[] | null;
  skills: Skill[] | null;
}

export enum Status {
  Active,
  Deleted,
  Expired,
  Inactive,
}

export enum VerificationType {
  FileUpload,
  Picture,
  Location,
  VoiceNote,
}

export enum VerificationMethod {
  Manual,
  Automatic,
}

export enum OrganizationStatus {
  Inactive,
  Active,
  Declined,
  Deleted,
}

export enum OrganizationDocumentType {
  Registration,
  EducationProvider,
  Business,
}

export enum OrganizationProviderType {
  Education,
  Marketplace,
}
export interface OpportunityCategory {
  id: string;
  name: string;
}

export interface OpportunityVerificationType {
  id: string;
  type?: VerificationType;
  displayName: string;
  description: string;
}

export interface OpportunityRequestBase {
  id: string | null;
  title: string;
  description: string;
  typeId: string;
  organizationId: string;
  instructions: string | null;
  uRL: string | null;
  zltoReward: number | null;
  yomaReward: number | null;
  zltoRewardPool: number | null;
  yomaRewardPool: number | null;
  verificationEnabled: boolean | null;
  verificationMethod: VerificationMethod | null | string;
  difficultyId: string;
  commitmentIntervalId: string;
  commitmentIntervalCount: number | null;
  participantLimit: number | null;
  keywords: string[] | null;
  dateStart: string | null;
  dateEnd: string | null;
  credentialIssuanceEnabled: boolean;
  ssiSchemaName: string | null;
  categories: string[];
  countries: string[];
  languages: string[];
  skills: string[];
  verificationTypes: OpportunityVerificationType[] | null;
  postAsActive: boolean;
}

export interface OpportunityCategory {
  id: string;
  name: string;
}
export interface OpportunityDifficulty {
  id: string;
  name: string;
}

export interface OpportunityType {
  id: string;
  name: string;
}
